src/software/jetson_nano/thunderloop.cpp

```diff
@@ -35,37 +91,98 @@ Thunderloop::~Thunderloop()
 
 /*
  * Run the main robot loop!
- *
- * @param The rate to run the loop
  */
-void Thunderloop::run(unsigned run_at_hz)
+void Thunderloop::runLoop()
 {
-    using clock = std::chrono::steady_clock;
+    // Timing
+    struct timespec next_shot;
+    struct timespec start_time;
+    struct timespec end_time;
+    struct timespec time;
+
+#define TIME_EXPRESSION(arg)                                                             \
+    clock_gettime(0, &start_time);                                                       \
+    arg;                                                                                 \
+    clock_gettime(0, &end_time);                                                         \
+    timespec_diff(&end_time, &start_time, &time);
 
-    auto next_frame = clock::now();
+    // Input buffer
+    TbotsProto::PrimitiveSet new_primitive_set;
+    TbotsProto::Vision new_vision;
 
+    // Loop interval
+    int interval = 1 / loop_hz_ * static_cast<int>(NANOSECONDS_PER_SECOND);
+
+    // Start the services
     motor_service_->start();
 
+    // Get current time
+    clock_gettime(0, &next_shot);
+
+    // Start after one second
+    next_shot.tv_sec++;
+
     for (;;)
     {
-        // TODO (#2335) add loop timing introspection and use Preempt-RT (maybe)
-        next_frame += std::chrono::milliseconds(
-            static_cast<int>(MILLISECONDS_PER_SECOND / run_at_hz));
+        // Wait until next shot
+        clock_nanosleep(0, TIMER_ABSTIME, &next_shot, NULL);
+
+        // Poll network service and grab most recent messages
+        TIME_EXPRESSION(std::tie(new_primitive_set, new_vision) =
```

Comment:
why do we need a tie?

Reply:
we don't _need_ it per say I was looking for `x, y = somefunction_that_returns_a_tuple()` equivalent in C++